kafka-ui-api/src/main/java/com/provectus/kafka/ui/serde/schemaregistry/SchemaRegistryRecordSerDe.java

```diff
@@ -0,0 +1,213 @@
+    }
+
+  }
+
+  private MessageFormatter getMessageFormatter(ConsumerRecord<Bytes, Bytes> msg, boolean isKey) {
+    if (isKey) {
+      return keyFormatMap.computeIfAbsent(msg.topic(), k -> detectFormat(msg, true));
+    } else {
+      return valueFormatMap.computeIfAbsent(msg.topic(), k -> detectFormat(msg, false));
+    }
+  }
+
+  private MessageFormatter detectFormat(ConsumerRecord<Bytes, Bytes> msg, boolean isKey) {
+    if (schemaRegistryClient != null) {
+      try {
+        final Optional<String> type = getSchemaFromMessage(msg, isKey)
+            .or(() -> getSchemaBySubject(msg.topic(), isKey).map(Schema::getSchemaType));
+        if (type.isPresent()) {
+          if (type.get().equals(MessageFormat.PROTOBUF.name())) {
+            try {
+              protobufFormatter.format(msg.topic(), msg.value().get());
+              return protobufFormatter;
+            } catch (Throwable e) {
+              log.info("Failed to get Protobuf schema for topic {}", msg.topic(), e);
+            }
+          } else if (type.get().equals(MessageFormat.AVRO.name())) {
+            try {
+              avroFormatter.format(msg.topic(), msg.value().get());
+              return avroFormatter;
+            } catch (Throwable e) {
+              log.info("Failed to get Avro schema for topic {}", msg.topic(), e);
+            }
+          } else if (type.get().equals(MessageFormat.JSON.name())) {
+            try {
+              jsonFormatter.format(msg.topic(), msg.value().get());
+              return jsonFormatter;
+            } catch (Throwable e) {
+              log.info("Failed to parse json from topic {}", msg.topic());
+            }
+          }
+        }
+      } catch (Exception e) {
+        log.warn("Failed to get Schema for topic {}", msg.topic(), e);
+      }
+    }
+
+    try {
+      jsonFormatter.format(msg.topic(), msg.value().get());
+      return jsonFormatter;
+    } catch (Throwable e) {
+      log.info("Failed to parse json from topic {}", msg.topic());
+    }
+
+    return stringFormatter;
+  }
+
+  @SneakyThrows
+  private Optional<String> getSchemaFromMessage(ConsumerRecord<Bytes, Bytes> msg, boolean isKey) {
+    Optional<String> result = Optional.empty();
+    final Bytes value = isKey ? msg.key() : msg.value();
+    if (value != null) {
+      ByteBuffer buffer = ByteBuffer.wrap(value.get());
+      if (buffer.get() == 0) {
+        int id = buffer.getInt();
+        result = Optional.ofNullable(
+            schemaRegistryClient.getSchemaById(id)
+        ).map(ParsedSchema::schemaType);
+      }
+    }
+    return result;
+  }
+
+  @SneakyThrows
+  private Optional<Schema> getSchemaBySubject(String topic, boolean isKey) {
+
+    String schemaName = String.format(
+        isKey ? cluster.getKeySchemaNameTemplate()
+            : cluster.getSchemaNameTemplate(), topic
+    );
+
+    final List<Integer> versions = schemaRegistryClient.getAllVersions(schemaName);
+    if (!versions.isEmpty()) {
```

Comment:
Can we use `var metadata = schemaRegistryClient.getLatestSchemaMetadata()` and then `metadata.getSchema()`?